File: packages/client-generator/src/runtime/send.ts
import { abortError } from './errors.js';
import { defaultRetryOn, retryDelay, sleep } from './retry.js';
import type {
  ClientConfig,
  Middleware,
  OperationContext,
  RequestContext,
  RequestOptions,
  RetryConfig,
} from './types.js';

/**
 * Optional behaviors the send core can use but never statically imports — wired by
 * `createClient` (the same seam the future inline-mode assembler relies on).
 */
export type SendCapabilities = {
  /** Serialize a typed multipart body (a plain object) to FormData. */
  serializeMultipart?: (body: Record<string, unknown>) => FormData;
};

/**
 * Normalize a caller's `HeadersInit` (plain record, `Headers` instance, or entry pairs)
 * to a plain record — spreading a `Headers` or an array contributes no entries.
 */
export function toHeaderRecord(headers: HeadersInit | undefined): Record<string, string> {
  if (headers === undefined) return {};
  if (headers instanceof Headers) {
    const record: Record<string, string> = {};
    headers.forEach((value, key) => {
      record[key] = value;
    });
    return record;
  }
  if (Array.isArray(headers)) return Object.fromEntries(headers);
  return headers;
}

/**
 * The effective middleware chain for a request: the single `onRequest`/`onResponse`/
 * `onError` config hooks as one implicit first middleware, then `config.middleware`.
 */
export function middlewareChain(config: ClientConfig): Middleware[] {
  const single =
    config.onRequest || config.onResponse || config.onError
      ? [{ onRequest: config.onRequest, onResponse: config.onResponse, onError: config.onError }]
      : [];
  return [...single, ...(config.middleware ?? [])];
}

/**
 * The fetch core shared by every operation: default + config + per-call headers, the
 * `onRequest` chain (BEFORE body serialization, so mutations are sent), body
 * serialization (JSON, or FormData via the multipart capability), the retry loop
 * (idempotent-only defaults, `Retry-After`, abandoned-body drain), and the reverse
 * `onResponse` onion. Returns the final response plus the request context.
 */
export async function send(
  config: ClientConfig,
  op: OperationContext,
  url: string,
  init: RequestOptions,
  body: unknown | undefined,
  bodySpec: { contentType: string; multipart?: boolean } | undefined,
  caps: SendCapabilities,
  accept = 'application/json'
): Promise<{ response: Response; context: RequestContext }> {
  const { retry: callRetry, timeout: callTimeout, idempotencyKey: callKey, ...fetchInit } = init;
  const retry: RetryConfig = { ...config.retry, ...callRetry };
  const timeout = callTimeout ?? config.timeout;
  const idempotency = callKey ?? config.idempotencyKey;
  const extra = typeof config.headers === 'function' ? await config.headers() : config.headers;
  const headers: Record<string, string> = {
    Accept: accept,
    ...extra,
    ...toHeaderRecord(fetchInit.headers),
  };
  const method = (fetchInit.method ?? 'GET').toUpperCase();
  // One stable key per LOGICAL call — set before the retry loop so every attempt
  // re-sends the same key; a caller-provided header always wins.
  if (
    idempotency !== undefined &&
    idempotency !== false &&
    (method === 'POST' || method === 'PATCH') &&
    !('Idempotency-Key' in headers) &&
    !('idempotency-key' in headers)
  ) {
    headers['Idempotency-Key'] =
      typeof idempotency === 'string'
        ? idempotency
        : typeof idempotency === 'function'
          ? idempotency()
          : crypto.randomUUID();
  }
  // Client identification for the API owner's telemetry — never in browsers, where a
  // custom header would force a CORS preflight the API may not allow.
  if (
    typeof config.clientHeader === 'string' &&
    typeof document === 'undefined' &&
    !('X-Redocly-Client' in headers) &&
    !('x-redocly-client' in headers)
  ) {
    headers['X-Redocly-Client'] = config.clientHeader;
  }
  const context: RequestContext = {
    url,
    method: fetchInit.method ?? 'GET',
    headers,
    body,
    operation: op,
  };
  const middleware = middlewareChain(config);
  for (const mw of middleware) if (mw.onRequest) await mw.onRequest(context);
  // Serialize AFTER onRequest so body mutations (case conversion, enveloping, signing) take effect.
  let payload: BodyInit | undefined;
  if (context.body !== undefined) {
    const value = context.body;
    const isBinary =
      value instanceof Blob ||
      value instanceof ArrayBuffer ||
      ArrayBuffer.isView(value as ArrayBufferView);
    const isFormData = typeof FormData !== 'undefined' && value instanceof FormData;
    const isURLSearchParams = value instanceof URLSearchParams;
    if (isFormData || isURLSearchParams || isBinary || typeof value === 'string') {
      payload = value as BodyInit;
    } else if (bodySpec?.multipart === true) {
      if (!caps.serializeMultipart) {
        throw new Error('Multipart capability not wired: cannot serialize the request body');
      }
      payload = caps.serializeMultipart(value as Record<string, unknown>);
    } else {
      payload = JSON.stringify(value);
      if (!('Content-Type' in context.headers) && !('content-type' in context.headers)) {
        // The spec's declared request content type (e.g. application/merge-patch+json).
        context.headers['Content-Type'] = bodySpec?.contentType ?? 'application/json';
      }
    }
  }
  const doFetch = config.fetch ?? fetch;
  const maxAttempts = 1 + (retry.retries ?? 0);
  const retryOn = retry.retryOn ?? defaultRetryOn;
  const signal = fetchInit.signal ?? undefined;

  let attempt = 0;
  while (true) {
    attempt++;
    if (signal?.aborted) throw abortError(signal);
    // A fresh timeout budget per attempt; the caller's signal still wins the race.
    // The composed signal also governs reading the response body.
    const attemptSignal = timeout
      ? signal
        ? AbortSignal.any([signal, AbortSignal.timeout(timeout)])
        : AbortSignal.timeout(timeout)
      : signal;
    let response: Response;
    try {
      response = await doFetch(context.url, {
        ...fetchInit,
        signal: attemptSignal,
        method: context.method,
        headers: context.headers,
        body: payload,
      });
    } catch (error) {
      if (
        attempt < maxAttempts &&
        !signal?.aborted &&
        (await retryOn({ attempt, request: context, error }))
      ) {
        await sleep(retryDelay(retry, attempt, null), signal);
        continue;
      }
      throw error;
    }
    // Reverse order: the last-registered middleware wraps closest to the network (onion).
    for (let i = middleware.length - 1; i >= 0; i--) {
      const onResponse = middleware[i].onResponse;
      if (onResponse) {
        const replaced = await onResponse(response, context);
        if (replaced && replaced !== response) {
          // Cancel the abandoned original's body — like the retry path, an unread body
          // keeps its connection checked out under Node/undici.
          await response.body?.cancel().catch(() => undefined);
          response = replaced;
        }
      }
    }
    if (
      !response.ok &&
      attempt < maxAttempts &&
      !signal?.aborted &&
      (await retryOn({ attempt, request: context, response }))
    ) {
      const retryAfter = response.headers.get('retry-after');
      // Drain the abandoned response body before the next attempt: an unread body
      // keeps the connection checked out (and can stall the pool) under Node/undici
      // and other strict HTTP clients. Ignore errors (e.g. a middleware already read it).
      await response.body?.cancel().catch(() => undefined);
      await sleep(retryDelay(retry, attempt, retryAfter), signal);
      continue;
    }
    return { response, context };
  }
}

File: packages/client-generator/src/generators/python/runtime/_send.py
# The request core for generated Python clients — mirror of the TypeScript
# runtime's send.ts: default + config + per-call headers, on_request middleware
# BEFORE serialization (mutations are sent), the retry loop (idempotent-methods
# default, Idempotency-Key opt-in makes POST/PATCH safe, Retry-After honored,
# exponential backoff with full jitter, a fresh timeout budget per attempt), and
# the reverse on_response onion.
from __future__ import annotations

import asyncio
import random
import time
import uuid
from dataclasses import dataclass
from typing import Any, Dict, Generic, List, Optional, Tuple, TypeVar

import httpx

from ._errors import ApiTimeoutError

T = TypeVar("T")


@dataclass
class Envelope(Generic[T]):
    """A *_with_headers() result: decoded body + coerced declared headers + raw response."""

    data: T
    headers: Dict[str, Any]
    response: httpx.Response


def read_envelope_headers(
    response: httpx.Response, specs: List[Tuple[str, str, str]]
) -> Dict[str, Any]:
    """Coerce declared response headers per (name, key, type) specs; absent/unparsable omitted."""
    headers: Dict[str, Any] = {}
    for name, key, type_ in specs:
        raw = response.headers.get(name)
        if raw is None:
            continue
        if type_ in ("integer", "number"):
            try:
                headers[key] = int(raw) if type_ == "integer" else float(raw)
            except ValueError:
                pass
        elif type_ == "boolean":
            lower = raw.strip().lower()
            if lower in ("true", "false"):
                headers[key] = lower == "true"
        else:
            headers[key] = raw
    return headers


_IDEMPOTENT_METHODS = {"GET", "HEAD", "PUT", "DELETE", "OPTIONS"}
_TRANSIENT_STATUS = {408, 429, 500, 502, 503, 504}


def _default_retry_on(method: str, headers: Dict[str, str], response: Optional[httpx.Response]) -> bool:
    safe = method.upper() in _IDEMPOTENT_METHODS or "Idempotency-Key" in headers
    if not safe:
        return False
    return response is None or response.status_code in _TRANSIENT_STATUS


def _retry_delay(retry: Dict[str, Any], attempt: int, retry_after: Optional[str]) -> float:
    if retry_after:
        try:
            return float(retry_after)
        except ValueError:
            pass  # HTTP-date form: fall through to backoff
    base = float(retry.get("retry_delay", 1.0))
    raw = base if retry.get("retry_strategy") == "fixed" else base * (2 ** (attempt - 1))
    return random.uniform(0, raw) if retry.get("jitter", True) is not False else raw


def send(
    client: httpx.Client,
    config: Dict[str, Any],
    op: Dict[str, Any],
    url: str,
    *,
    method: str,
    headers: Optional[Dict[str, str]] = None,
    params: Optional[Dict[str, Any]] = None,
    json_body: Any = None,
    content: Any = None,
    data: Any = None,
    files: Any = None,
    timeout: Optional[float] = None,
    idempotency_key: Any = None,
    retry: Optional[Dict[str, Any]] = None,
) -> httpx.Response:
    merged_retry: Dict[str, Any] = {**(config.get("retry") or {}), **(retry or {})}
    effective_timeout = timeout if timeout is not None else config.get("timeout")
    merged_headers: Dict[str, str] = {**(config.get("headers") or {}), **(headers or {})}

    # One stable key per LOGICAL call — set before the retry loop so every
    # attempt re-sends the same key; a caller-provided header always wins.
    key = idempotency_key if idempotency_key is not None else config.get("idempotency_key")
    if (
        key not in (None, False)
        and method.upper() in ("POST", "PATCH")
        and "Idempotency-Key" not in merged_headers
    ):
        merged_headers["Idempotency-Key"] = (
            key if isinstance(key, str) else key() if callable(key) else str(uuid.uuid4())
        )

    context = {
        "url": url,
        "method": method.upper(),
        "headers": merged_headers,
        "body": json_body,
        "operation": op,
    }
    middleware: List[Any] = config.get("middleware") or []
    for mw in middleware:
        on_request = getattr(mw, "on_request", None) or (mw.get("on_request") if isinstance(mw, dict) else None)
        if on_request:
            on_request(context)

    max_attempts = 1 + int(merged_retry.get("retries", 0))
    retry_on = merged_retry.get("retry_on") or (
        lambda ctx: _default_retry_on(context["method"], context["headers"], ctx.get("response"))
    )

    attempt = 0
    while True:
        attempt += 1
        try:
            response = client.request(
                context["method"],
                context["url"],
                headers=context["headers"],
                params=params,
                json=context["body"] if content is None and files is None and data is None else None,
                content=content,
                data=data,
                files=files,
                timeout=effective_timeout if effective_timeout is not None else httpx.USE_CLIENT_DEFAULT,
            )
        except httpx.TimeoutException:
            if attempt < max_attempts and retry_on({"attempt": attempt, "response": None}):
                time.sleep(_retry_delay(merged_retry, attempt, None))
                continue
            raise ApiTimeoutError(op.get("id", "?"), float(effective_timeout or 0), attempt) from None
        except httpx.TransportError:
            if attempt < max_attempts and retry_on({"attempt": attempt, "response": None}):
                time.sleep(_retry_delay(merged_retry, attempt, None))
                continue
            raise

        for mw in reversed(middleware):
            on_response = getattr(mw, "on_response", None) or (mw.get("on_response") if isinstance(mw, dict) else None)
            if on_response:
                replaced = on_response(response, context)
                if replaced is not None:
                    response = replaced

        if (
            not response.is_success
            and attempt < max_attempts
            and retry_on({"attempt": attempt, "response": response})
        ):
            time.sleep(_retry_delay(merged_retry, attempt, response.headers.get("retry-after")))
            continue
        return response


async def send_async(
    client: httpx.AsyncClient,
    config: Dict[str, Any],
    op: Dict[str, Any],
    url: str,
    *,
    method: str,
    headers: Optional[Dict[str, str]] = None,
    params: Optional[Dict[str, Any]] = None,
    json_body: Any = None,
    content: Any = None,
    data: Any = None,
    files: Any = None,
    timeout: Optional[float] = None,
    idempotency_key: Any = None,
    retry: Optional[Dict[str, Any]] = None,
) -> httpx.Response:
    """The async mirror of send() — same retry/timeout/idempotency semantics."""
    merged_retry: Dict[str, Any] = {**(config.get("retry") or {}), **(retry or {})}
    effective_timeout = timeout if timeout is not None else config.get("timeout")
    merged_headers: Dict[str, str] = {**(config.get("headers") or {}), **(headers or {})}
    key = idempotency_key if idempotency_key is not None else config.get("idempotency_key")
    if (
        key not in (None, False)
        and method.upper() in ("POST", "PATCH")
        and "Idempotency-Key" not in merged_headers
    ):
        merged_headers["Idempotency-Key"] = (
            key if isinstance(key, str) else key() if callable(key) else str(uuid.uuid4())
        )
    context = {
        "url": url,
        "method": method.upper(),
        "headers": merged_headers,
        "body": json_body,
        "operation": op,
    }
    middleware: List[Any] = config.get("middleware") or []
    for mw in middleware:
        on_request = getattr(mw, "on_request", None) or (mw.get("on_request") if isinstance(mw, dict) else None)
        if on_request:
            on_request(context)
    max_attempts = 1 + int(merged_retry.get("retries", 0))
    retry_on = merged_retry.get("retry_on") or (
        lambda ctx: _default_retry_on(context["method"], context["headers"], ctx.get("response"))
    )
    attempt = 0
    while True:
        attempt += 1
        try:
            response = await client.request(
                context["method"],
                context["url"],
                headers=context["headers"],
                params=params,
                json=context["body"] if content is None and files is None and data is None else None,
                content=content,
                data=data,
                files=files,
                timeout=effective_timeout if effective_timeout is not None else httpx.USE_CLIENT_DEFAULT,
            )
        except httpx.TimeoutException:
            if attempt < max_attempts and retry_on({"attempt": attempt, "response": None}):
                await asyncio.sleep(_retry_delay(merged_retry, attempt, None))
                continue
            raise ApiTimeoutError(op.get("id", "?"), float(effective_timeout or 0), attempt) from None
        except httpx.TransportError:
            if attempt < max_attempts and retry_on({"attempt": attempt, "response": None}):
                await asyncio.sleep(_retry_delay(merged_retry, attempt, None))
                continue
            raise
        for mw in reversed(middleware):
            on_response = getattr(mw, "on_response", None) or (mw.get("on_response") if isinstance(mw, dict) else None)
            if on_response:
                replaced = on_response(response, context)
                if replaced is not None:
                    response = replaced
        if (
            not response.is_success
            and attempt < max_attempts
            and retry_on({"attempt": attempt, "response": response})
        ):
            await asyncio.sleep(_retry_delay(merged_retry, attempt, response.headers.get("retry-after")))
            continue
        return response
